components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -199,127 +318,237 @@ impl RangeManager {
             return Err(FailedReason::TooOldRead);
         }
 
-        meta.range_snapshot_list.new_snapshot(read_ts);
-        Ok(meta.id)
+        meta.region_snapshot_list.new_snapshot(read_ts);
+        Ok(())
     }
 
     // If the snapshot is the last one in the snapshot list of one cache range in
-    // historical_ranges, it means one or some evicted_ranges may be ready to be
+    // historical_regions, it means one or some evicted_ranges may be ready to be
     // removed physically.
     // So, we return a vector of ranges to denote the ranges that are ready to be
-    // removed.
-    pub(crate) fn remove_range_snapshot(
+    // removed.range_key
+    pub(crate) fn remove_region_snapshot(
         &mut self,
         snapshot_meta: &RangeCacheSnapshotMeta,
-    ) -> Vec<CacheRange> {
-        if let Some(range_key) = self
-            .historical_ranges
-            .iter()
-            .find(|&(range, meta)| {
-                range.contains_range(&snapshot_meta.range) && meta.id == snapshot_meta.range_id
-            })
-            .map(|(r, _)| r.clone())
+    ) -> Vec<Region> {
+        // fast path: in most case, region is change, so we change region first.
+        if let Some(region_meta) = self.regions.get_mut(&snapshot_meta.region_id)
+            && region_meta.region.get_region_epoch().version == snapshot_meta.epoch_version
         {
-            let meta = self.historical_ranges.get_mut(&range_key).unwrap();
-            meta.range_snapshot_list
+            // epoch not changed
+            region_meta
+                .region_snapshot_list
                 .remove_snapshot(snapshot_meta.snapshot_ts);
-            if meta.range_snapshot_list.is_empty() {
-                self.historical_ranges.remove(&range_key);
+            if Self::region_ready_to_evict(region_meta, &self.historical_regions) {
+                region_meta.set_state(RegionState::Evicting);
+                return vec![region_meta.region.clone()];
             }
+            return vec![];
+        }
 
-            return self
-                .ranges_being_deleted
-                .keys()
-                .filter(|evicted_range| {
-                    !self
-                        .historical_ranges
-                        .keys()
-                        .any(|r| r.overlaps(evicted_range))
-                })
-                .cloned()
-                .collect::<Vec<_>>();
-        }
-
-        // It must belong to the `self.ranges` if not found in `self.historical_ranges`
-        let range_key = self
-            .ranges
-            .iter()
-            .find(|&(range, meta)| {
-                range.contains_range(&snapshot_meta.range) && meta.id == snapshot_meta.range_id
-            })
-            .map(|(r, _)| r.clone())
-            .unwrap();
-        let meta = self.ranges.get_mut(&range_key).unwrap();
-        meta.range_snapshot_list
+        // slow path: region not found or epoch version changes, must fell in the
```

Comment:
Are you saying every time change the epoch of a region, must move the old one to the historical_regions?

Reply:
only when region epoch version change and there are active snapshots. Actually, this only happens for region split because we current just evict the region for region merge.